sunscreen_math/src/metal_impl/ristrettovec.rs

```diff
@@ -0,0 +1,716 @@
+        self.len * size_of::<RistrettoPoint>()
+    }
+
+    pub fn get(&self, index: usize) -> RistrettoPoint {
+        if index > self.len {
+            panic!("Index {index} exceeds bounds of {}", self.len);
+        }
+
+        let mut x = [0u32; 10];
+        let mut y = [0u32; 10];
+        let mut z = [0u32; 10];
+        let mut t = [0u32; 10];
+
+        let buffer_slice = self.buffer_slice();
+
+        for i in 0..10 {
+            x[i] = buffer_slice[i * self.len + index];
+        }
+
+        for i in 0..10 {
+            y[i] = buffer_slice[(i + 10) * self.len + index];
+        }
+
+        for i in 0..10 {
+            z[i] = buffer_slice[(i + 20) * self.len + index];
+        }
+
+        for i in 0..10 {
+            t[i] = buffer_slice[(i + 30) * self.len + index];
+        }
+
+        RistrettoPoint(EdwardsPoint {
+            X: CannonicalFieldElement(x).to_field(),
+            Y: CannonicalFieldElement(y).to_field(),
+            Z: CannonicalFieldElement(z).to_field(),
+            T: CannonicalFieldElement(t).to_field(),
+        })
+    }
+
+    pub fn iter(&self) -> RistrettoPoints {
+        RistrettoPoints {
+            vec: self,
+            i: 0usize,
+        }
+    }
+
+    // TODO: This probably needs to return a slice of MaybeUninit<u32>.
+    fn buffer_slice_mut(&mut self) -> &mut [u32] {
+        let byte_len = self.len * size_of::<RistrettoPoint>();
+
+        unsafe { slice::from_raw_parts_mut(self.data.contents() as *mut u32, byte_len) }
+    }
+
+    fn buffer_slice(&self) -> &[u32] {
+        let byte_len = self.len * size_of::<RistrettoPoint>();
+
+        unsafe { slice::from_raw_parts(self.data.contents() as *const u32, byte_len) }
+    }
+}
+
+impl Add<RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn add(self, rhs: RistrettoPointVec) -> Self::Output {
+        &self + &rhs
+    }
+}
+
+impl Add<&RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn add(self, rhs: &RistrettoPointVec) -> Self::Output {
+        &self + rhs
+    }
+}
+
+impl Add<RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn add(self, rhs: RistrettoPointVec) -> Self::Output {
+        self + &rhs
+    }
+}
+
+impl Add<&RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn add(self, rhs: &RistrettoPointVec) -> Self::Output {
+        assert_eq!(self.len(), rhs.len());
+
+        let runtime = Runtime::get();
+
+        let len = self.len_bytes();
+
+        let o = Self::Output {
+            data: runtime.alloc(len),
+            len: rhs.len(),
+        };
+
+        let len_gpu = U32Arg::new(rhs.len() as u32);
+
+        // TODO: o gets mutated here. Need to figure out what that means in terms of UB.
+        runtime.run(
+            "ristretto_add",
+            &[&self.data, &rhs.data, &o.data, &len_gpu.data],
+            [(rhs.len() as u64, 64), (1, 1), (1, 1)],
```

Comment:
What does "in terms of UB" mean? Upper bound?

Reply:
UB = undefined behavior. It's a bad scary thing you don't want.